File: utils/wind_statistics.py
"""
wind_statistics.py
Конвертация MATLAB wind_statistics.m → Python

Определяет преобладающее направление ветра по компонентам U, V
и возвращает диапазон скорости (центральные 80% данных).
"""

import numpy as np


def wind_statistics(u_wind: np.ndarray, v_wind: np.ndarray) -> tuple[str, float, float]:
    """
    Рассчитывает преобладающее направление и диапазон скорости ветра.

    Parameters
    ----------
    u_wind : np.ndarray  — зональная компонента ветра (м/с)
    v_wind : np.ndarray  — меридиональная компонента ветра (м/с)

    Returns
    -------
    final_dir   : str   — текстовое описание направления ветра
    s_min       : float — нижняя граница скорости (10-й перцентиль)
    s_max       : float — верхняя граница скорости (90-й перцентиль)
    """
    # Модуль скорости
    S = np.sqrt(u_wind**2 + v_wind**2)
    S_vec = S.flatten()

    # Отсекаем крайние 10% с обеих сторон
    v_min_threshold = np.nanquantile(S_vec, 0.10)
    v_max_threshold = np.nanquantile(S_vec, 0.90)

    # Маска центральных 80% данных
    mask = (S >= v_min_threshold) & (S <= v_max_threshold)

    U_typ = u_wind[mask]
    V_typ = v_wind[mask]

    # Перевод в метеорологические румбы (откуда дует ветер, 0–360°, 0 = Север)
    angles = np.mod(180 + np.degrees(np.arctan2(U_typ, V_typ)), 360)

    # Разбивка на 8 румбов по 45°
    r_idx = (np.floor(np.mod(angles + 22.5, 360) / 45)).astype(int)  # 0..7

    # Статистика направлений
    counts = np.bincount(r_idx, minlength=8)
    percents = counts / len(r_idx) * 100
    sorted_r = np.argsort(percents)[::-1]   # индексы по убыванию %
    sorted_p = percents[sorted_r]

    dirs_text = [
        "северный", "северо-восточный", "восточный", "юго-восточный",
        "южный", "юго-западный", "западный", "северо-западный"
    ]

    # Логика определения направления (полностью идентична MATLAB-оригиналу)
    if sorted_p[0] >= 35 and sorted_p[0] >= sorted_p[1] * 1.5:
        final_dir = dirs_text[sorted_r[0]]

    elif (sorted_p[0] + sorted_p[1]) >= 50:
        final_dir = f"{dirs_text[sorted_r[0]]} и {dirs_text[sorted_r[1]]}"

    elif (sorted_p[0] + sorted_p[1] + sorted_p[2]) >= 65:
        final_dir = (
            f"{dirs_text[sorted_r[0]]}, "
            f"{dirs_text[sorted_r[1]]} и "
            f"{dirs_text[sorted_r[2]]}"
        )

    else:
        final_dir = "переменных направлений"

    s_min = round(float(v_min_threshold), 1)
    s_max = round(float(v_max_threshold), 1)

    return final_dir, s_min, s_max

File: matlab_original/wind_statistics.m
function [final_dir, s_min, s_max]=wind_statistics(U_wind, V_wind)
% Исходные данные для примера (замените на свои матрицы 37x25)
% U, V — компоненты ветра

% Расчет модуля скорости
S = sqrt(U_wind.^2 + V_wind.^2);
S_vec = S(:);

% отсекаем крайние 10% с обеих сторон
v_min_threshold = quantile(S_vec, 0.10);
v_max_threshold = quantile(S_vec, 0.90);

% Создаем маску центральные 80% данных
mask = (S >= v_min_threshold) & (S <= v_max_threshold);

% Извлекаем данные только для типичных условий
U_typ = U_wind(mask);
V_typ = V_wind(mask);
S_typ = S(mask);

% перевод в метеорологические румбы (откуда дует)
% Формула перевода U,V в градусы (0-360, где 0 - Север)
angles = mod(180 + atan2d(U_typ, V_typ), 360);

% Разбивка на 8 румбов (каждый по 45 градусов, центр первого - 0)
% 1-С, 2-СВ, 3-В, 4-ЮВ, 5-Ю, 6-ЮЗ, 7-З, 8-СЗ
r_idx = floor(mod(angles + 22.5, 360) / 45) + 1;

% Статистика распределения направлений в процентах
counts = histcounts(r_idx, 1:9);
percents = (counts / numel(r_idx)) * 100;
[sorted_p, sorted_r] = sort(percents, 'descend'); % Сортировка по убыванию %

% Формирование текстового описания направления
dirs_text = {'северный', 'северо-восточный', 'восточный', 'юго-восточный', ...
             'южный', 'юго-западный', 'западный', 'северо-западный'};

% Если один румб > 35% И он существенно (в 1.5 раза) выше второго места
if sorted_p(1) >= 35 && (sorted_p(1) >= sorted_p(2) * 1.5)
    final_dir = [dirs_text{sorted_r(1)}];
    
% Если сумма двух первых > 50% и они сопоставимы
elseif (sorted_p(1) + sorted_p(2)) >= 50
    final_dir = [dirs_text{sorted_r(1)}, ' и ', dirs_text{sorted_r(2)}];
    
% Если сумма трех первых > 65%
elseif (sorted_p(1) + sorted_p(2) + sorted_p(3)) >= 65
    final_dir = [dirs_text{sorted_r(1)}, ', ', dirs_text{sorted_r(2)}, ...
                 ' и ', dirs_text{sorted_r(3)}];
    
% Если данные "размазаны" ровным слоем (лидер меньше 20-25%)
else
    final_dir = 'переменных направлений';
end

% Итоговая строка для отчета
s_min = round(v_min_threshold,1);
s_max = round(v_max_threshold,1);

end